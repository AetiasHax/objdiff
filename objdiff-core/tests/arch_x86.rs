use objdiff_core::{diff, obj};

mod common;

#[test]
#[cfg(feature = "x86")]
fn read_x86() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/staticdebug.obj"), &diff_config).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "?PrintThing@@YAXXZ").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_combine_sections() {
    let diff_config = diff::DiffObjConfig {
        combine_data_sections: true,
        combine_text_sections: true,
        ..Default::default()
    };
    let obj = obj::read::parse(include_object!("data/x86/rtest.obj"), &diff_config).unwrap();
    insta::assert_debug_snapshot!(obj.sections);
}
