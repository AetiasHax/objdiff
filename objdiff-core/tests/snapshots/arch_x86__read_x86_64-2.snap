---
source: objdiff-core/tests/arch_x86.rs
expression: diff.instruction_rows
---
[
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 0,
                size: 5,
                opcode: 414,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 5,
                size: 5,
                opcode: 414,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 10,
                size: 1,
                opcode: 640,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 11,
                size: 4,
                opcode: 740,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 15,
                size: 5,
                opcode: 414,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 20,
                size: 5,
                opcode: 414,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 25,
                size: 4,
                opcode: 448,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 29,
                size: 4,
                opcode: 460,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 33,
                size: 5,
                opcode: 414,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 38,
                size: 5,
                opcode: 414,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 43,
                size: 5,
                opcode: 448,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 48,
                size: 5,
                opcode: 460,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 53,
                size: 4,
                opcode: 11,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 57,
                size: 5,
                opcode: 414,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 62,
                size: 5,
                opcode: 414,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 67,
                size: 5,
                opcode: 448,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 72,
                size: 5,
                opcode: 460,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 77,
                size: 4,
                opcode: 11,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 81,
                size: 4,
                opcode: 7,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 85,
                size: 1,
                opcode: 590,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 86,
                size: 1,
                opcode: 662,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
]
